contrib/hamilton/contrib/user/skrawcz/fine_tuning/__init__.py

```diff
@@ -0,0 +1,653 @@
+    :param max_target_length:
+    :return: the prediction, and the ground truth label.
+    """
+    # generate summary
+    # outputs = model.generate(
+    # input_ids=sample["input_ids"].unsqueeze(0).cuda(),
+    # do_sample=True, top_p=0.9, max_new_tokens=max_target_length)
+    outputs = model.generate(
+        input_ids=sample["input_ids"].unsqueeze(0).to(device),
+        do_sample=True,
+        top_p=0.9,
+        max_new_tokens=max_target_length,
+    )
+    prediction = tokenizer.decode(outputs[0].detach().cpu().numpy(), skip_special_tokens=True)
+    # decode eval sample
+    # Replace -100 in the labels as we can't decode them.
+    labels = np.where(sample["labels"] != -100, sample["labels"], tokenizer.pad_token_id)
+
+    # Some simple post-processing
+    labels = tokenizer.decode(labels, skip_special_tokens=True)
+    return prediction, labels
+
+
+# run predictions
+# this can take ~45 minutes
+@extract_fields({"validation_predictions": list, "validation_references": list})
+def finetuned_model_on_validation_set(
+    tokenized_validation: Dataset,
+    finetuned_model: torch.nn.Module,
+    tokenizer: PreTrainedTokenizerBase,
+    device: torch.device,
+) -> dict[str, list]:
+    """Evaluates our fine-tuned model on the validation set.
+
+    If you run this on a large model this can take a while.
+    :param tokenized_validation:
+    :param finetuned_model:
+    :param tokenizer:
+    :param device:
+    :return:
+    """
+    predictions, references = [], []
+    with torch.inference_mode():
+        for sample in tqdm(tokenized_validation.with_format("torch")):
+            prediction, labels = _evaluate_peft_model(
+                sample=sample,
+                model=finetuned_model,
+                tokenizer=tokenizer,
+                max_target_length=512,
+                device=device,
+            )
+            predictions.append(prediction)
+            references.append(labels)
+            print("#" * 10)
+            print(f"prediction = {prediction}, labels = {labels}")
+    return {"validation_predictions": predictions, "validation_references": references}
+
+
+# compute metric
+def validation_set_metrics(
+    validation_predictions: list,
+    validation_references: list,
+    metric: evaluate.EvaluationModule,
+) -> dict:
+    rogue = metric.compute(
+        predictions=validation_predictions,
+        references=validation_references,
+        use_stemmer=True,
+    )
+    eval_rouge_scores = {
+        "rogue1": rogue["rouge1"] * 100,
+        "rouge2": rogue["rouge2"] * 100,
+        "rougeL": rogue["rougeL"] * 100,
+        "rougeLsum": rogue["rougeLsum"] * 100,
+    }
+    return eval_rouge_scores
+
+
+def training_and_validation_set_metrics(
+    validation_set_metrics: dict, fit_trainer_eval_metrics: dict
+) -> dict:
+    """Combines the training and validation set metrics."""
+    return {**fit_trainer_eval_metrics, **validation_set_metrics}
+
+
+def inference_set_predictions(
```

Comment:
This should really be a second pipeline or an external script?

Reply:
no. this is hamilton -- run what you want.

Reply:
Disagree, I think this should be a different one. This does too much.